File: scripts/github/extract_pr_code_comments.ts

import { Octokit } from '@octokit/rest';
// Environment variables provided by shell or runtime

// Configuration
const GITHUB_TOKEN = process.env.GH_TOKEN_CORE_GITHUB_WORKER;

if (!GITHUB_TOKEN) {
    console.error("Error: GH_TOKEN_CORE_GITHUB_WORKER environment variable is not set.");
    console.error("Please export it: export GH_TOKEN_CORE_GITHUB_WORKER='your_token_here'");
    process.exit(1);
}

const octokit = new Octokit({
    auth: GITHUB_TOKEN,
});

function parsePrUrl(url: string) {
    try {
        const u = new URL(url);
        const pathParts = u.pathname.split('/').filter(Boolean);

        // Expected: /owner/repo/pull/number
        if (pathParts.length < 4 || pathParts[2] !== 'pull') {
            throw new Error('Invalid GitHub PR URL');
        }

        return {
            owner: pathParts[0],
            repo: pathParts[1],
            pull_number: parseInt(pathParts[3], 10)
        };
    } catch (e: any) {
        throw new Error(`Invalid GitHub PR URL. Expected format: https://github.com/owner/repo/pull/number. ${e.message}`);
    }
}

const extractSuggestion = (body: string): string | undefined => {
    const pattern = /```suggestion\r?\n([\s\S]*?)\r?\n```/;
    const match = body.match(pattern);
    return match ? match[1] : undefined;
}

async function fetchAllReviewComments(owner: string, repo: string, pull_number: number) {
    const comments: any[] = [];

    // Octokit automatic pagination
    const iterator = octokit.paginate.iterator(octokit.rest.pulls.listReviewComments, {
        owner,
        repo,
        pull_number,
        per_page: 100,
    });

    for await (const { data } of iterator) {
        comments.push(...data);
    }

    return comments;
}

function transformToSchema(comment: any) {
    return {
        id: comment.id,
        path: comment.path,
        line: comment.line,
        start_line: comment.start_line,
        original_line: comment.original_line,
        body: comment.body || "",
        diff_hunk: comment.diff_hunk,
        suggestion: extractSuggestion(comment.body || ""),
        user: {
            login: comment.user?.login || "Unknown",
            avatar_url: comment.user?.avatar_url || ""
        },
        created_at: comment.created_at,
        html_url: comment.html_url
    };
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.error("Usage: npx ts-node tests/extract_pr_code_comments.ts <PR_URL>");
        process.exit(1);
    }

    const prUrl = args[0];

    try {
        const { owner, repo, pull_number } = parsePrUrl(prUrl);
        // console.error(`Fetching comments for ${owner}/${repo} PR #${pull_number}...`);

        const rawComments = await fetchAllReviewComments(owner, repo, pull_number);
        const formattedComments = rawComments.map(transformToSchema);

        console.log(JSON.stringify(formattedComments, null, 2));

    } catch (e: any) {
        console.error(`Error: ${e.message}`);
        process.exit(1);
    }
}

main();

File: scripts/github/extract_pr_code_comments.py
import os
import sys
import re
import json
import requests
from urllib.parse import urlparse

def get_github_token():
    """Retrieves the GitHub token from the environment variable."""
    token = os.getenv("MY_GITHUB_TOKEN")
    if not token:
        print("Error: MY_GITHUB_TOKEN environment variable is not set.", file=sys.stderr)
        sys.exit(1)
    return token

def parse_pr_url(url):
    """Parses the GitHub PR URL to extract owner, repo, and pull_number."""
    parsed_url = urlparse(url)
    path_parts = parsed_url.path.strip("/").split("/")
    
    if len(path_parts) < 4 or path_parts[2] != "pull":
        print("Error: Invalid GitHub Pull Request URL.", file=sys.stderr)
        print("Expected format: https://github.com/OWNER/REPO/pull/NUMBER", file=sys.stderr)
        sys.exit(1)
        
    return {
        "owner": path_parts[0],
        "repo": path_parts[1],
        "pull_number": path_parts[3]
    }

def extract_suggestion(body):
    """Attempts to extract a code suggestion block from the comment body."""
    # GitHub suggestions are wrapped in ```suggestion ... ``` blocks
    pattern = r"```suggestion\r?\n(.*?)\r?\n```"
    match = re.search(pattern, body, re.DOTALL)
    if match:
        return match.group(1)
    return None

def fetch_comments(owner, repo, pull_number, token):
    """Fetches review comments from the GitHub API, handling pagination."""
    url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pull_number}/comments"
    headers = {
        "Accept": "application/vnd.github+json",
        "Authorization": f"Bearer {token}",
        "X-GitHub-Api-Version": "2022-11-28"
    }
    
    all_comments = []
    page = 1
    
    while True:
        response = requests.get(url, headers=headers, params={"page": page, "per_page": 100})
        
        if response.status_code != 200:
            print(f"Error fetching comments: {response.status_code}", file=sys.stderr)
            print(response.text, file=sys.stderr)
            sys.exit(1)
            
        data = response.json()
        if not data:
            break
            
        all_comments.extend(data)
        page += 1
        
    return all_comments

def format_comment(comment):
    """Maps a GitHub API comment object to the requested JSON schema."""
    return {
        "id": comment.get("id"),
        "path": comment.get("path"),
        "line": comment.get("line"), # Nullable in API, maps to Zod nullable
        "start_line": comment.get("start_line"), # Optional/Nullable
        "original_line": comment.get("original_line"), # Useful for outdated comments
        "body": comment.get("body"),
        "diff_hunk": comment.get("diff_hunk"),
        "suggestion": extract_suggestion(comment.get("body", "")),
        "user": {
            "login": comment.get("user", {}).get("login"),
            "avatar_url": comment.get("user", {}).get("avatar_url")
        },
        "created_at": comment.get("created_at"),
        "html_url": comment.get("html_url")
    }

def main():
    if len(sys.argv) != 2:
        print("Usage: python3 extract_comments.py <PR_URL>", file=sys.stderr)
        sys.exit(1)

    pr_url = sys.argv[1]
    token = get_github_token()
    pr_details = parse_pr_url(pr_url)
    
    print(f"Fetching comments for {pr_details['owner']}/{pr_details['repo']} PR #{pr_details['pull_number']}...", file=sys.stderr)
    
    raw_comments = fetch_comments(pr_details['owner'], pr_details['repo'], pr_details['pull_number'], token)
    formatted_comments = [format_comment(c) for c in raw_comments]
    
    # Print JSON to stdout
    print(json.dumps(formatted_comments, indent=2))

if __name__ == "__main__":
    main()